File: data/derived_cot/rq1_traces/translation/r1/java_to_py/code_output/EmailClient.py
import datetime

class EmailClient:
    def __init__(self, addr, capacity):
        self.addr = addr
        self.capacity = capacity
        self.inbox = []
    
    def sendTo(self, recv, content, size):
        if not recv.isFullWithOneMoreEmail(size):
            timestamp = datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S")
            email = {
                "sender": self.addr,
                "receiver": recv.addr,
                "content": content,
                "size": size,
                "time": timestamp,
                "state": "unread"
            }
            recv.inbox.append(email)
            return True
        else:
            self.clearInbox(size)
            if recv.isFullWithOneMoreEmail(size):
                return False
            return self.sendTo(recv, content, size)
    
    def fetch(self):
        if not self.inbox:
            return None
        for email in self.inbox:
            if email["state"] == "unread":
                email["state"] = "read"
                return email
        return None
    
    def isFullWithOneMoreEmail(self, size):
        occupiedSize = self.getOccupiedSize()
        return occupiedSize + size > self.capacity
    
    def getOccupiedSize(self):
        occupiedSize = 0.0
        for email in self.inbox:
            size_val = email["size"]
            occupiedSize += float(size_val)
        return occupiedSize
    
    def clearInbox(self, size):
        if not self.addr:
            return
        freedSpace = 0.0
        while freedSpace < size and self.inbox:
            email = self.inbox.pop(0)
            freedSpace += float(email["size"])
    
    def getInbox(self):
        return self.inbox
    
    def setInbox(self, inbox):
        self.inbox = inbox

File: data/derived_cot/rq1_traces/translation/r1/java_to_cpp/code_output/EmailClient.cpp
#include <deque>
#include <unordered_map>
#include <string>
#include <any>
#include <chrono>
#include <ctime>
#include <sstream>
#include <iostream>
#include <iomanip>

#if defined(_WIN32)
#define NOMINMAX
#include <Windows.h>
#endif

std::string getCurrentTimestamp() {
    auto now = std::chrono::system_clock::now();
    std::time_t t = std::chrono::system_clock::to_time_t(now);
    std::tm tm;
#if defined(_WIN32)
    localtime_s(&tm, &t);
#else
    localtime_r(&t, &tm);
#endif
    char buffer[80];
    std::strftime(buffer, sizeof(buffer), "%Y-%m-%d %H:%M:%S", &tm);
    return std::string(buffer);
}

class EmailClient {
private:
    std::string addr;
    double capacity;
    std::deque<std::unordered_map<std::string, std::any>> inbox;

public:
    EmailClient(const std::string& addr, double capacity) : addr(addr), capacity(capacity), inbox() {}

    bool sendTo(EmailClient& recv, const std::string& content, double size) {
        if (!recv.isFullWithOneMoreEmail(size)) {
            std::string timestamp = getCurrentTimestamp();
            std::unordered_map<std::string, std::any> email;
            email["sender"] = this->addr;
            email["receiver"] = recv.addr;
            email["content"] = content;
            email["size"] = size;
            email["time"] = timestamp;
            email["state"] = std::string("unread");
            recv.inbox.push_back(email);
            return true;
        } else {
            this->clearInbox(size);
            if (recv.isFullWithOneMoreEmail(size)) {
                return false;
            }
            return sendTo(recv, content, size);
        }
    }

    std::unordered_map<std::string, std::any>* fetch() {
        if (inbox.empty()) {
            return nullptr;
        }
        for (auto& email : inbox) {
            auto stateIt = email.find("state");
            if (stateIt != email.end()) {
                try {
                    std::string state = std::any_cast<std::string>(stateIt->second);
                    if (state == "unread") {
                        email["state"] = std::string("read");
                        return &email;
                    }
                } catch (const std::bad_any_cast&) {
                    continue;
                }
            }
        }
        return nullptr;
    }

    bool isFullWithOneMoreEmail(double size) {
        double occupiedSize = getOccupiedSize();
        return occupiedSize + size > capacity;
    }

    double getOccupiedSize() {
        double occupiedSize = 0;
        for (const auto& email : inbox) {
            auto it = email.find("size");
            if (it != email.end()) {
                try {
                    double emailSize = std::any_cast<double>(it->second);
                    occupiedSize += emailSize;
                } catch (const std::bad_any_cast&) {
                }
            }
        }
        return occupiedSize;
    }

    void clearInbox(double size) {
        if (addr.empty()) {
            return;
        }
        double freedSpace = 0;
        while (freedSpace < size && !inbox.empty()) {
            auto email = inbox.front();
            inbox.pop_front();
            auto it = email.find("size");
            if (it != email.end()) {
                try {
                    double emailSize = std::any_cast<double>(it->second);
                    freedSpace += emailSize;
                } catch (const std::bad_any_cast&) {
                }
            }
        }
    }

    std::deque<std::unordered_map<std::string, std::any>>& getInbox() {
        return inbox;
    }

    void setInbox(const std::deque<std::unordered_map<std::string, std::any>>& newInbox) {
        inbox = newInbox;
    }
};